File: Commons/Geometry/src/main/java/com/greentree/commons/geometry/geom2d/AABB.java
package com.greentree.commons.geometry.geom2d;

import com.greentree.commons.geometry.math.float2;

/** @author Arseny Latyshev */
public class AABB {

	private final static float BORDER_COEFFICIENT = 0.0000001f;

	private final float2 min, max;

	public AABB(float minx, float miny, float w, float h) {
		this(new float2(minx, miny), new float2(minx + w, miny + h));
	}

	public AABB(float2...fs) {
		min = new float2();
		max = new float2();

		max.x = -Float.MAX_VALUE;
		max.y = -Float.MAX_VALUE;
		min.x = Float.MAX_VALUE;
		min.y = Float.MAX_VALUE;

		for(var p : fs) {
			max.x = Math.max(max.x, p.x);
			max.y = Math.max(max.y, p.y);
			min.x = Math.min(min.x, p.x);
			min.y = Math.min(min.y, p.y);
		}
		expand(min, max);
	}

	public AABB(IShape2D shape) {
		min = new float2();
		max = new float2();
		max.x = -Float.MAX_VALUE;
		max.y = -Float.MAX_VALUE;
		min.x = Float.MAX_VALUE;
		min.y = Float.MAX_VALUE;
		for(var p : shape.getPoints()) {
			max.x = Math.max(max.x, p.x());
			max.y = Math.max(max.y, p.y());
			min.x = Math.min(min.x, p.x());
			min.y = Math.min(min.y, p.y());
		}
		expand(min, max);
	}

	private static void expand(float2 min, float2 max) {
		float dd = (max.x - min.x + max.y - min.y + 1) * BORDER_COEFFICIENT;
		min.y -= dd;
		max.y += dd;
		min.x -= dd;
		max.x += dd;
	}

	public float getHeight() {
		return max.y - min.y;
	}

	public float getMaxX() {
		return max.x;
	}

	public float getMaxY() {
		return max.y;
	}

	public float getMediunX() {
		return (max.x + min.x) / 2f;
	}

	public float getMediunY() {
		return (max.y + min.y) / 2f;
	}

	public float getMinX() {
		return min.x;
	}

	public float getMinY() {
		return min.y;
	}

	public float getWidth() {
		return max.x - min.x;
	}

	public boolean isIntersect(AABB other) {
		return max.x >= other.min.x && min.x <= other.max.x && max.y >= other.min.y && min.y <= other.max.y;
	}

	public boolean isIntersect(float x, float y) {
		return max.x >= x && min.x <= x && max.y >= y && min.y <= y;
	}

	@Override
	public String toString() {
		return "AABB [min=" + min + ", max=" + max + "]";
	}

}

File: Commons/Geometry/src/main/java/com/greentree/commons/geometry/geom2d/AABB.kt
package com.greentree.commons.geometry.geom2d

import com.greentree.commons.math.vector.AbstractVector2f
import com.greentree.commons.math.vector.minus
import kotlin.math.max
import kotlin.math.min

/** @author Arseny Latyshev
 */
class AABB {

	val minX: Float
	val maxX: Float
	val minY: Float
	val maxY: Float

	constructor(vararg points: AbstractVector2f) {
		var maxx = -Float.MAX_VALUE
		var maxy = -Float.MAX_VALUE
		var minx = Float.MAX_VALUE
		var miny = Float.MAX_VALUE
		for(p in points) {
			maxx = max(maxx, p.x)
			maxy = max(maxy, p.y)
			minx = min(minx, p.x)
			miny = min(miny, p.y)
		}
		val sX = (maxx - minx) * BORDER_COEFFICIENT / 2
		minx += sX
		maxx -= sX
		val sY = (maxy - miny) * BORDER_COEFFICIENT / 2
		miny += sY
		maxy -= sY
		minX = minx
		maxX = maxx
		minY = miny
		maxY = maxy
	}

	constructor(
		minX: Float,
		maxX: Float,
		minY: Float,
		maxY: Float,
	) {
		var minx = minX
		var maxx = maxX
		var miny = minY
		var maxy = maxY
		val sX = (maxx - minx) * BORDER_COEFFICIENT / 2
		minx += sX
		maxx -= sX
		val sY = (maxy - miny) * BORDER_COEFFICIENT / 2
		miny += sY
		maxy -= sY
		this.minX = minx
		this.maxX = maxx
		this.minY = miny
		this.maxY = maxy
	}

	val height: Float
		get() = maxY - minY
	val centerX: Float
		get() = (maxX + minX) / 2f
	val centerY: Float
		get() = (maxY + minY) / 2f
	val width: Float
		get() = maxX - minX

	fun isIntersect(other: AABB): Boolean {
		return maxX >= other.minX && minX <= other.maxX && maxY >= other.minY && minY <= other.maxY
	}

	fun isInside(x: Float, y: Float): Boolean {
		return minX <= x && x <= maxX && minY <= y && y <= maxY
	}

	override fun toString(): String {
		return "AABB(minX=$minX, maxX=$maxX, minY=$minY, maxY=$maxY)"
	}

	companion object {

		private const val BORDER_COEFFICIENT = 0.000001f
	}
}